File: HW/HW 10/Solution.py
# Christopher Oh
# SJSU CS 146
# HW 10 - Binary Search Tree Level Order Traversal
from typing import Optional, List
import TreeNode

class Solution:
    def __init__(self):
        print("HW 10")

    def levelOrder(self, root) -> List[List[int]]:
        nodes_by_level = []

        if root is not None:
            self.helper(root, nodes_by_level, 0)

        return nodes_by_level

    def helper(self, node, nodes_by_level, level):
        while len(nodes_by_level) - 1 < level:
            nodes_by_level.append([])

        nodes_by_level[level].append(node.val)

        if node.left is not None:
            self.helper(node.left, nodes_by_level, level + 1)

        if node.right is not None:
            self.helper(node.right, nodes_by_level, level + 1)

File: HW/HW 10/Solution.java
/**
 * Christopher Oh
 * SJSU CS 146
 * HW 10 - BST level order tree traversal
 */

import java.util.ArrayList;
import java.util.List;

public class Solution {
    public Solution () {}


    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> list = new ArrayList<>();

        if (root != null) helper(root, list, 0);

        return list;
    }

    /**
     * Adds the node to its level in the tree
     * The index of the arraylist in list is its level
     * ex) list.get(3) refers to level 3 on the tree
     * root is at level 0
     * @param node
     * @param list
     * @param level
     */
    private void helper (TreeNode node, List<List<Integer>> list, int level) {
        /**
         * if the list doesn't have as many levels as the node in the parameter, add levels to list until it does
         */
        while (list.size() - 1 < level) {
            ArrayList<Integer> temp = new ArrayList<>();
            list.add(temp);
        }

        list.get(level).add(node.val);
        if (node.left != null) helper(node.left, list, level + 1);
        if (node.right != null) helper(node.right, list, level + 1);
    }

}
